src/utilities/include/par.h

```diff
@@ -49,47 +53,251 @@ inline constexpr ExecutionPolicy autoPolicy(size_t size) {
-    switch (policy) {                                         \
-      case ExecutionPolicy::Par:                              \
-        return std::NAME(std::execution::par_unseq, args...); \
-      case ExecutionPolicy::Seq:                              \
-        break;                                                \
-    }                                                         \
-    return std::NAME(args...);                                \
+template <typename Iter, typename F>
+void for_each(ExecutionPolicy policy, Iter first, Iter last, F f) {
+#if MANIFOLD_PAR == 'T'
+  if (policy == ExecutionPolicy::Par) {
+    tbb::parallel_for(tbb::blocked_range<Iter>(first, last),
+                      [&f](const tbb::blocked_range<Iter> &range) {
+                        for (Iter i = range.begin(); i != range.end(); i++)
+                          f(*i);
+                      });
+    return;
   }
-#define STL_DYNAMIC_BACKEND_VOID(NAME)                 \
-  template <typename... Args>                          \
-  void NAME(ExecutionPolicy policy, Args... args) {    \
-    switch (policy) {                                  \
-      case ExecutionPolicy::Par:                       \
-        std::NAME(std::execution::par_unseq, args...); \
-        break;                                         \
-      case ExecutionPolicy::Seq:                       \
-        std::NAME(args...);                            \
-        break;                                         \
-    }                                                  \
+#endif
+  std::for_each(first, last, f);
+}
+
+template <typename Iter, typename F>
+void for_each_n(ExecutionPolicy policy, Iter first, size_t n, F f) {
+  for_each(policy, first, first + n, f);
+}
+
+template <typename InputIter, typename BinaryOp,
+          typename T = typename std::iterator_traits<InputIter>::value_type>
+T reduce(ExecutionPolicy policy, InputIter first, InputIter last, T init,
+         BinaryOp f) {
+#if MANIFOLD_PAR == 'T'
+  if (policy == ExecutionPolicy::Par) {
+    // should we use deterministic reduce here?
+    return tbb::parallel_reduce(
+        tbb::blocked_range<InputIter>(first, last), init,
+        [&f](const tbb::blocked_range<InputIter> &range, T value) {
+          for (InputIter i = range.begin(); i != range.end(); i++)
+            value = f(value, *i);
+          return value;
+        },
+        f);
   }
-#else
-#define STL_DYNAMIC_BACKEND(NAME, RET)             \
-  template <typename Ret = RET, typename... Args>  \
-  Ret NAME(ExecutionPolicy policy, Args... args) { \
-    return std::NAME(args...);                     \
+#endif
+  return std::reduce(first, last, init, f);
+}
+
+template <typename InputIter, typename BinaryOp, typename UnaryOp,
+          typename T = std::invoke_result_t<
+              UnaryOp, typename std::iterator_traits<InputIter>::value_type>>
+T transform_reduce(ExecutionPolicy policy, InputIter first, InputIter last,
```

Comment:
Should we just build our auto execution policy into these wrappers so we don't need to define an `autoPolicy` all the time? We can just use e.g. `std::reduce` if we want to force a serial operation.

Reply:
Yes, we can do that. This is usually just `std::distance`.

Reply:
Thinking about it, we may still want execution policy for things like `for_each`, where each loop may be doing something non-trivial and will benefit from parallelization even when the number of iteration is small. Will think about how to find out all of those...

Reply:
but we can have an overload that automatically figure out the policy.

Reply:
The benchmark I added may help this as well. I think the threshold may be quite different for each algorithm.